course3/homework-string/my_string.cc

```diff
@@ -0,0 +1,192 @@
+#include "my_string.h"
+
+#include <cstring>
+#include <memory>
+
+MyString::MyString() 
+	:
+	mBuffer(nullptr)
+{
+}
+
+MyString::MyString(const MyString& that)
+{
+	if (that.empty())
+	{
+		mBuffer = nullptr;
+	}
+	else
+	{
+		mBuffer = std::make_unique<char[]>(std::strlen(that.mBuffer.get()) + 1);
+		std::strcpy(mBuffer.get(), that.mBuffer.get());
+	}
+}
+
+MyString::MyString(MyString&& that)
+{
+	mBuffer.swap(that.mBuffer);
+}
+
+MyString::MyString(const char* str)
+{
+	if (str == nullptr || str[0] == '\0')
+	{
+		mBuffer = nullptr;
+	}
+	else
+	{
+		mBuffer = std::make_unique<char[]>(std::strlen(str) + 1);
+		std::strcpy(mBuffer.get(), str);
+	}
+}
+
+MyString& MyString::operator=(const MyString& that)
+{
+	if (this == &that)
+		return *this;
+	
+	if (that.empty())
+	{
+		std::unique_ptr<char[]> oldBuffer = nullptr;
+		mBuffer.swap(oldBuffer);
+	}
+	else
+	{
+		auto newBuffer = std::make_unique<char[]>(std::strlen(that.mBuffer.get()) + 1);
+		std::strcpy(newBuffer.get(), that.mBuffer.get());
+
+		mBuffer.swap(newBuffer);
+	}
+
+	return *this;
+}
+
+MyString& MyString::operator=(MyString&& that)
+{
+	if (this == &that)
+		return *this;
+
+	mBuffer.swap(that.mBuffer);
+
+	return *this;
+}
+
+MyString MyString::operator+(const MyString& that) const
+{
+	auto thisSize = size();
+	auto thatSize = that.size();
+	
+	if (thisSize == 0 && thatSize == 0)
+	{
+		return MyString();
+	}
+	else if (thatSize == 0)
+	{
+		return MyString(mBuffer.get());
```

Comment:
But this function returns MyString by copy, so copy will happen.
Remember - every type here is a "value type", every return value and parameter is passed "by value"